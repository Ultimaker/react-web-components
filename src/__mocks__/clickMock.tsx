// Copyright (c) 2018 Ultimaker B.V.

export const mockClickEvent = { stopPropagation: jest.fn(), preventDefault: jest.fn() };
