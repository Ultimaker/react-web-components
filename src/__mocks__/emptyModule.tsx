// Copyright (c) 2018 Ultimaker B.V.
module.exports = {};
