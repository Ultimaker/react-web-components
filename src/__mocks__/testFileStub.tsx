// Copyright (c) 2018 Ultimaker B.V.
module.exports = 'test-file-stub';