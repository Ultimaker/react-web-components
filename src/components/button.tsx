import * as React from 'react';
import classNames from 'classnames';
import { Link } from 'react-router-dom';

// components
import Spinner from './spinner';

export type ButtonType = 'submit' | 'button' | 'link';
export type ButtonStyle = 'primary' | 'secondary' | 'quiet' | 'alert';
export type ButtonShape = 'rectangle' | 'circle' | 'pill';

export interface ButtonProps {
    /** Optional ID for the button **/
    id?: string;
    /** Called when the Button is clicked */
    onClickHandler?: () => void;
    /** Disables the button when true */
    disabled?: boolean;
    /** html button type: 'submit' | 'button' | 'link' */
    type?: ButtonType;
    /** CSS styling: 'primary' | 'secondary' | 'quiet' | 'alert' */
    style?: ButtonStyle;
    /** Visual shape of the Button: 'rectangle' | 'circle' | 'pill' */
    shape?: ButtonShape;
    /** Replaces the Button text with a spinner when true */
    showSpinner?: boolean;
    /** URL to link to for type link */
    linkURL?: string;
    /** If true button of type link will open in a new tab */
    linkToNewTab?: boolean;
    /** Additional classes for styling */
    className?: string;
}

export class Button extends React.Component<ButtonProps, {}> {

    static defaultProps = {
        type: 'button',
        style: 'primary',
        shape: 'rectangle',
        linkToNewTab: false,
        className: ''
    };

    constructor(props) {
        super(props);

        this._onClickHandler = this._onClickHandler.bind(this);
    }

    private _isLinkInternal(): boolean {
        const { linkURL, linkToNewTab } = this.props;
        return !linkToNewTab && linkURL && !(/^https?:\/\//.test(linkURL));
    }

    private _onClickHandler(e: React.MouseEvent<HTMLElement>): void {
        const { onClickHandler, disabled, showSpinner } = this.props;

        e.stopPropagation();
        if (onClickHandler && !(disabled || showSpinner)) {
            onClickHandler();
        }
    }

    private _renderInternalLink(classes: string): JSX.Element {
        const { id, disabled, showSpinner, linkURL, children } = this.props;

        return <Link
            id={id}
            className={classes}
            to={disabled || showSpinner ? undefined : linkURL}
        >
            <span className="text">{children}</span>
        </Link>
    }

    private _renderExternalLink(classes: string): JSX.Element {
        const { id, disabled, showSpinner, linkURL, linkToNewTab, children } = this.props;

        return <a 
            id={id} 
            className={classes}
            href={disabled || showSpinner ? undefined : linkURL}
            target={linkToNewTab ? '_blank' : undefined}
        >
            <span className="text">{children}</span>
            {showSpinner &&
                <Spinner />
            }
        </a>
    }

    private _renderButton(classes: string): JSX.Element {
        const { id, disabled, showSpinner, type, children } = this.props;

        return <button id={id} className={classes} onClick={this._onClickHandler} disabled={disabled || showSpinner} type={type}>
            <span className="text">{children}</span>
            {showSpinner &&
                <Spinner />
            }
        </button>
    }

    render(): JSX.Element {
        const { disabled, type, style, shape, showSpinner, className } = this.props;

        const classes = classNames(`btn btn--${style} btn--${shape} ${className}`,
            { disabled }, { waiting: showSpinner });

        if (type === 'link' && this._isLinkInternal()) {
            return this._renderInternalLink(classes);
        }
        else if (type === 'link') {
            return this._renderExternalLink(classes);
        }
        else {
            return this._renderButton(classes);
        }
    }
}

export default Button;
