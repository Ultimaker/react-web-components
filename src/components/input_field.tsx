import * as React from 'react';

import {InputFieldProps} from './input_fields/input_field_wrapper'
import NumberField from './input_fields/number_field'
import TextareaField from './input_fields/textarea_field'
import SelectField from './input_fields/select_field'
import CheckboxField from './input_fields/checkbox_field'
import ImageField from './input_fields/image_field'
import DateField from './input_fields/date_field'
import FileField from './input_fields/file_field'
import ChildrenField from './input_fields/children_field'
import TagsField from './input_fields/tags_field'
import TextField from './input_fields/text_field'

export interface OldInputFieldProps extends InputFieldProps {
    type?: 'text' | 'number' | 'textarea' | 'password' | 'email' | 'url' | 'select' | 'checkbox' |
        'image' | 'date' | 'file' | 'tags' | 'children';
}

const input_fields = {
    number: NumberField,
    textarea: TextareaField,
    select: SelectField,
    checkbox: CheckboxField,
    image: ImageField,
    date: DateField,
    file: FileField,
    children: ChildrenField,
    tags: TagsField,
}

/**
 * @deprecated - Use the separate components
 */
const InputField: React.StatelessComponent<OldInputFieldProps> = ({type, ...props}) => {
    if (type === 'text' || type === 'password' || type === 'email' || type === 'url') {
        return <TextField type={type} {...props} />;
    } else {
        let Component = input_fields[type];
        return <Component {...props} />;
    }
};

InputField.displayName = "InputField";
InputField.defaultProps = {
    type: 'text',
    labelLayoutWidth: '1/1',
    labelWidthBreakpoint: 'sm'
};

export default InputField;
