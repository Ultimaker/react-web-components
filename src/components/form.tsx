import * as React from 'react';
import { Link } from 'react-router-dom';

import { default as Button, ButtonStyle } from './button';


/** The validation of each field may be a text, a list of elements or the validation of a sub-model **/
export type FormValidationResponse = { [key: string]: string | JSX.Element[] | FormValidationResponse };

export interface FormProps {
	/** Primary button text */
	primaryBtnText?: string;
	/** Called when the primary button is clicked to submit the form */
	onSubmitHandler: () => void;
	/** Primary button style type */
	primaryBtnStyle?: ButtonStyle;
	/** Secondary button text */
	secondaryBtnText?: string;
	/** Called when the secondary button is clicked */
	secondaryBtnHandler?: () => void;
	/** Primary button style type */
	secondaryBtnStyle?: ButtonStyle;
	/** A url to link to instead of calling secondaryBtnHandler */
	secondaryBtnLink?: string;
	/** The form validation state and validation error messages */
	validationErrors?: FormValidationResponse;
	/** Override the form validation and enable the primary button */
	alwaysEnableSubmitButton?: boolean;
}

export interface FormState {
	primaryBtnSpinner: boolean;
	secondaryBtnSpinner: boolean;
	submitted: boolean;
}


export class Form extends React.Component<FormProps, FormState> {

	state = {
		primaryBtnSpinner: false,
		secondaryBtnSpinner: false,
		submitted: false
	}

	constructor(props) {
		super(props);

		// bind callbacks once
		this._onSubmitHandler = this._onSubmitHandler.bind(this);
		this._secondaryBtnHandler = this._secondaryBtnHandler.bind(this);
		this._renderChild = this._renderChild.bind(this);
	}

	_onSubmitHandler(e: React.FormEvent<HTMLFormElement>): void {
		e.preventDefault();
		this.setState({ submitted: true });
		this.props.onSubmitHandler();
	}

	_secondaryBtnHandler(): void {
		this.props.secondaryBtnHandler();
	}

	/**
 * Renders a single child of the form component. If the child has the `id` props, we will check for errors in the
 * form validation, any errors are passed as extra props to the child.
	 * @param child - The child element to be rendered.
	 * @private
	 */
	private _renderChild(child: JSX.Element): JSX.Element {
		const { validationErrors } = this.props;
		const { submitted } = this.state;
		const errors = validationErrors && child && child.props && validationErrors[child.props.id];

		return child && (
			<div className="form__item">
				{React.cloneElement(child, errors && {
					validationError: errors,
					submitted
				})}
			</div>
		)
	}

	render(): JSX.Element {
		const { primaryBtnText, secondaryBtnText, primaryBtnStyle, secondaryBtnStyle, validationErrors,
			secondaryBtnLink, alwaysEnableSubmitButton, children } = this.props;
		const { primaryBtnSpinner, secondaryBtnSpinner } = this.state;

		return (
			<form noValidate className="form" onSubmit={this._onSubmitHandler}>
				{React.Children.map(children, this._renderChild)}
				{primaryBtnText &&
					<div className="form__actions">
						{primaryBtnText &&
							<div className="form__btn-container">
								<Button
									style={primaryBtnStyle}
									disabled={alwaysEnableSubmitButton ? false : secondaryBtnSpinner || validationErrors !== null}
									type="submit"
									showSpinner={primaryBtnSpinner}>

									{primaryBtnText}
								</Button>
							</div>
						}

						{secondaryBtnText && !secondaryBtnLink &&
							<div className="form__btn-container">
								<Button
									style={secondaryBtnStyle}
									disabled={primaryBtnSpinner}
									onClickHandler={this._secondaryBtnHandler}
									showSpinner={secondaryBtnSpinner}
								>
									{secondaryBtnText}
								</Button>
							</div>
						}

						{secondaryBtnText && secondaryBtnLink &&
							<div className="form__btn-container">
								<Link to={secondaryBtnLink} className="btn btn--quiet">
									<span className="text">{secondaryBtnText}</span>
								</Link>
							</div>
						}

					</div>
				}
			</form>
		)
	}
}

export default Form;