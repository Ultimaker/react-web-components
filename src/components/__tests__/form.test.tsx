// Copyright (c) 2018 Ultimaker B.V.
import 'jest';
import 'jsdom-global/register';
import * as React from 'react';
import { shallow, mount } from 'enzyme';

// component
import { Form } from '../form';
import { InputField } from '../input_field';
import { Button } from '../button';

// mocks
import { mockClickEvent } from '../../__mocks__/clickMock'

describe('The Form component', () => {
    let props;
    let wrapper;

    beforeEach(() => {
        props = {
            onSubmitHandler: jest.fn(),
            primaryBtnText: 'primaryBtnText'
        };
        wrapper = shallow(<Form {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
        expect(props.onSubmitHandler).not.toHaveBeenCalled();
    });

    it('should call onSubmitHandler when form is submitted', () => {
        wrapper.simulate('submit', mockClickEvent);
        expect(props.onSubmitHandler).toHaveBeenCalled();
        expect(wrapper.state().submitted).toEqual(true);
        expect(mockClickEvent.preventDefault).toBeCalled();
    });

    it('should render a secondary button', () => {
        wrapper.setProps({
            secondaryBtnText: 'secondaryBtnText',
            secondaryBtnHandler: jest.fn(),
            secondaryBtnStyle: 'secondary'
        });
        expect(props.onSubmitHandler).not.toHaveBeenCalled();
        expect(wrapper.find(Button)).toHaveLength(2);
    });

    it('should render with a form item', () => {
        const mountedWrapper = mount(
            <Form {...props}>
                <InputField id="test" onChangeHandler={jest.fn()} />
            </Form>
        );
        expect(mountedWrapper).toMatchSnapshot();
    });

    it('should pass validation error to form item', () => {
        const mountedWrapper = mount(
            <Form {...props} validationErrors={{ test: 'Validation error' }}>
                <InputField id="test" onChangeHandler={jest.fn()} />
            </Form>
        );
        expect(mountedWrapper.find(InputField).prop('validationError')).toEqual('Validation error');
        expect(mountedWrapper.find(Button).prop('disabled')).toEqual(true);
    });

});
