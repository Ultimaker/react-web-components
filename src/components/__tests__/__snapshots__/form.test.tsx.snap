// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Form component should render 1`] = `
<form
  className="form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="form__actions"
  >
    <div
      className="form__btn-container"
    >
      <Button
        className=""
        disabled={true}
        linkToNewTab={false}
        shape="rectangle"
        style="primary"
        type="submit"
      >
        primaryBtnText
      </Button>
    </div>
  </div>
</form>
`;

exports[`The Form component should render with a form item 1`] = `
<Form
  onSubmitHandler={[MockFunction]}
  primaryBtnText="primaryBtnText"
>
  <form
    className="form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form__item"
      key=".0"
    >
      <InputField
        id="test"
        labelLayoutWidth="1/1"
        labelWidthBreakpoint="sm"
        onChangeHandler={[MockFunction]}
        type="text"
      >
        <div
          className="input-field input-field--text layout"
        >
          <InputFieldInput
            id="test"
            labelLayoutWidth="1/1"
            labelWidthBreakpoint="sm"
            onChangeHandler={[Function]}
            type="text"
          >
            <div
              className="layout__item u-fill layout__item--middle"
            >
              <div
                className="input-container layout layout--gutter-xs"
              >
                <div
                  className="layout__item u-fill"
                >
                  <input
                    className="input"
                    id="test"
                    max={null}
                    min={null}
                    name="test"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </InputFieldInput>
        </div>
      </InputField>
    </div>
    <div
      className="form__actions"
    >
      <div
        className="form__btn-container"
      >
        <Button
          className=""
          disabled={true}
          linkToNewTab={false}
          shape="rectangle"
          style="primary"
          type="submit"
        >
          <button
            className="btn btn--primary btn--rectangle  disabled"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <span
              className="text"
            >
              primaryBtnText
            </span>
          </button>
        </Button>
      </div>
    </div>
  </form>
</Form>
`;
